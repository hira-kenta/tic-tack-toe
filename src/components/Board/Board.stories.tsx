import { Meta, StoryObj } from "@storybook/react";
import Board from "./Board";

const meta = {
    title: "tic-tac-toe/Board",
    component: Board,
    parameters: {
        layout: 'centered'
    },
    args:{
        xIsNext: true,
        squares: Array(9).fill('')
    },
    tags: ['autodocs'],
} satisfies Meta<typeof Board>

export default meta;
type Story = StoryObj<typeof Board>;

/**
 * 初期表示時
 */
export const Default: Story = {
    args:{
        xIsNext: true,
    }
}

/**
 * プレイ中
 */
export const Playing: Story = {
    args:{
        xIsNext: false,
        squares: ['X','O','X','O','X','','','','',],
    }
}

/**
 * X が勝利した場合の盤面
 */
export const WinningOfCross: Story = {
    args:{
        xIsNext: false,
        squares: ['X','O','X','O','X','O','X','','',],
    }
}

/**
 * O が勝利した場合の盤面
 */
export const WinningOfCircle: Story = {
    args:{
        xIsNext: false,
        squares: ['O','X','O','O','X','X','O','','X'],
    }
}

/**
 * すべてのマスに△がセットされた場合
 */
export const FilledWithTriangle: Story = {
    args:{
        xIsNext: false,
        squares: Array(9).fill('△'),
    }
}