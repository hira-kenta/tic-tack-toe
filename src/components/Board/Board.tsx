import React from "react";
import { calculateWinner } from "../../helpers/calculateWinner";
import Square from "../Square/Square";

type Props = {
  /** 次はXの番かを判定するフラグ */
  xIsNext: boolean,
  /** マスにセットされたOX。  
   * 初期表示時はすべて空文字がセットされている。
  */
  squares: string[],
  /**
   * クリックされたマスにOXをセットし、セットした位置情報を履歴に保存する処理
   * @param nextSquares 
   * @param location 
   * @returns 
   */
  onPlay: (nextSquares: string[], location: {row: number, column: number}) => void,
}

/**
 *  OXゲームの盤面
 *  - マスをクリックするとOかXが入力される
 *  - 次の手番と勝敗結果が表示される
 *  @returns 
 */
const Board = ({xIsNext, squares, onPlay}: Props) => {
    const handleClick = (i: number, row: number): void => {
      if(squares[i] || calculateWinner(squares)){
        return;
      }
      const location = setLocation(row, i);
      const nextSquares: string[] = squares.slice();
      nextSquares[i] = xIsNext ? 'X' : 'O';
      onPlay(nextSquares, location);
    }

    const setLocation = (row: number, column: number) => {
      let location: {row: number, column: number};
      const convertedRow = row + 1;
      const convertedColumn = (column % 3) + 1
      location = {row: convertedRow, column: convertedColumn}
      return location;
    }
  
    const winner = calculateWinner(squares);
    let status: string = "Draw...";
    if(winner){
      status = "Winner: " + winner.player;
    }else if(squares.includes('')){
      status = "次のプレイヤー: " + (xIsNext ? "X" : "O");
    }
  
    return (
    <>
      <div data-e2e="status" className="status">{status}</div>
      {Array(3).fill(null).map((_, row) => (
        <div className="board-row" key={row}>
          {Array(3).fill(null).map((_, col) => {
            const index = row * 3 + col;
            const isWinning = winner?.line.includes(index) ? true : false;
            return (
              <Square key={index} isWinning={isWinning} value={squares[index]} onSquareClick={() => handleClick(index, row)} />
            );
          })}
        </div>
      ))}
    </>
    );
  }

  export default Board;