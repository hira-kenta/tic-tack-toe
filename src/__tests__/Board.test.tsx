import React from 'react';
import renderer from 'react-test-renderer';
import * as stories from '../components/Board/Board.stories';
import { composeStories } from '@storybook/react';

const {Default: DefaultBoard, Playing: PlayingBoard, WinningOfCircle: WinningOfCircleBoard, WinningOfCross: WinningOfCrossBoard, FilledWithTriangle: FilledWithTriangleBoard} = composeStories(stories);

describe('Boardのスナップショットテスト', () => {
    test('初期表示', () => {
        const component = renderer.create(
            <DefaultBoard />
        )
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
    test('プレイ中', () => {
        const component = renderer.create(
            <PlayingBoard />
        )
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
    test('Xが勝利した場合の盤面', () => {
        const component = renderer.create(
            <WinningOfCrossBoard />
        )
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
    test('Oが勝利した場合の盤面', () => {
        const component = renderer.create(
            <WinningOfCircleBoard />
        )
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
    test('すべてのマスに△がセットされた場合', () => {
        const component = renderer.create(
            <FilledWithTriangleBoard />
        )
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
})