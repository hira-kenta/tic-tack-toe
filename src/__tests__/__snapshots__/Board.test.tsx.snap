// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boardのスナップショットテスト Oが勝利した場合の盤面 1`] = `
Array [
  <div
    className="status"
  >
    Winner: O
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square win"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      O
    </button>
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square win"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square win"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
  </div>,
]
`;

exports[`Boardのスナップショットテスト Xが勝利した場合の盤面 1`] = `
Array [
  <div
    className="status"
  >
    Winner: X
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="square win"
      onClick={[Function]}
    >
      X
    </button>
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square filled"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="square win"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      O
    </button>
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square win"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>,
]
`;

exports[`Boardのスナップショットテスト すべてのマスに△がセットされた場合 1`] = `
Array [
  <div
    className="status"
  >
    Draw...
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      △
    </button>
  </div>,
]
`;

exports[`Boardのスナップショットテスト プレイ中 1`] = `
Array [
  <div
    className="status"
  >
    次のプレイヤー: O
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square filled"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="square filled"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square"
      onClick={[Function]}
    />
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>,
]
`;

exports[`Boardのスナップショットテスト 初期表示 1`] = `
Array [
  <div
    className="status"
  >
    次のプレイヤー: X
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>,
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>,
]
`;
